pulsar/consumer_partition.go

```diff
@@ -1752,6 +1762,37 @@ func (pc *partitionConsumer) markScaleIfNeed() {
 	}
 }
 
+func (pc *partitionConsumer) reserveMemory(size int64) {
+	pc.client.memLimit.ForceReserveMemory(size)
+	if pc.client.memLimit.CurrentUsagePercent() >= receiverQueueShrinkMemThreshold {
```

Comment:
This shrinking strategy is different from java client implementation. 

The java client implementation:

Will try to trigger shrinking memory after each `ReserveMemory`, this means that the producer will also trigger the consumer to reduce the queue when sending data. 

https://github.com/shibd/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java#L65

Your implementation will cause the following scenario where producers cannot produce data.
- One client instance has `1` producer instance and `3` consumer.
- If `3` consumers run out of memory, the producer's production data will block.

Reply:
I think `shrink` revoked by `memory limiter` and revoked by `consumer` are equivalent.

> One client instance has 1 producer instance and 3 consumer.
> If 3 consumers run out of memory, the producer's production data will block.

In this scenario, the Java client will also block. Because `consumer` use the `forceReserve`.

By the way, is it necessary for the producer to trigger the consumer `shrink`? If the memory usage exceeds the threshold, the consumer will always shrink it's receiver queue, right?

Reply:
The reasons here are not the same as those implemented by the Java client: 
1. I don’t think it matters whether this is triggered by `memory limiter` as long as the `shrink` is not triggered repeated. 
3. If we want to implement it like the Java client, we need to modify the `memory limiter` to allow it to support a trigger, but I feel like this in itself is an unnecessary code coupling.

Reply:
> By the way, is it necessary for the producer to trigger the consumer shrink? If the memory usage exceeds the threshold, the consumer will always shrink it's receiver queue, right?

Sorry, the previous example is not quite accurate.

I think the key difference is whether one `consumer` or `producer` needs to trigger all consumers to reduce the queue size.

- The java client implementation:
When a producer or consumer call the `ReserveMemory` option, If the memory exceeds the threshold, it triggers all consumers to shrink the queue size

- Your implementation:
When a consumer checks that memory exceeds a threshold, it will only try to reduce its own queue size.

I think there is a memory usage starvation issue here:

- One client instance has `1` producer instance and `3` consumer.
- If the queue of these three consumers is already relatively large, if the user of `consumer-1` and `consumer 2` is not actively called (there are many messages piled up in queurCh), then if `consumer-3` checks and reduces its queue size after receiving a message, although the memory limit is reduced, but `consumer-1` and `consumer-2` still occupy more memory (a lot of pending consumption), 
- The same way, if the producer wants to send a message, Since it cannot trigger other consumers to reduce the queue size, it will also send more slowly.

In short, in this case, Memory is wasted on inactive consumers(consumer-1, consumer-2)

If it is in implementing the Java client, the problem does not exist. 
Each time, all the consumers will reduce the queue size, and then let them re-promote the  queue size according to the activity of the consumers.

Reply:
> In short, in this case, Memory is wasted on inactive consumers(consumer-1, consumer-2)

OK, I see. But even in Java client, `reduceReceiverQueue()` can not free memory immediately. It just controls next flow to prefetch fewer messages. IMO, the memory limit on the consumer side is lazy. The memory will only be freed when pending messages are consumed.

So I think this question can be summed up like this.
- Even `memory limiter` can trigger all the consumers to reduce it's queue size. The actual memory release still depends on the consumer's consumption.
- If the consumer is not active, even reducing the queue size will not reduce the memory.

Reply:
> OK, I see. But even in Java client, reduceReceiverQueue() can not free memory immediately. It just controls next flow to prefetch fewer messages.

> It just controls next flow to prefetch fewer messages.

Yes, We need these inactive consumers to reduce the size of the pull flow so that it can give memory to more active consumers.